core/network/nic.go

```diff
@@ -259,6 +243,58 @@ func (i *InterfaceInfo) PrimaryAddress() ProviderAddress {
 	return i.Addresses[0]
 }
 
+// InterfaceInfos is a slice of InterfaceInfo
+// for a single host/machine/container.
+type InterfaceInfos []InterfaceInfo
+
+// IterHierarchy runs the input function for every interface by processing each
+// device hierarchy, ensuring that no child device is processed before its
+// parent.
+func (s InterfaceInfos) IterHierarchy(f func(InterfaceInfo) error) error {
+	return s.iterChildHierarchy("", f)
+}
+
+func (s InterfaceInfos) iterChildHierarchy(parentName string, f func(InterfaceInfo) error) error {
+	children := s.Children(parentName)
+	for _, child := range children {
+		if err := f(child); err != nil {
+			return err
+		}
+		if err := s.iterChildHierarchy(child.InterfaceName, f); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Children returns interfaces that are direct children
+// of the interface with the input name.
+func (s InterfaceInfos) Children(parentName string) InterfaceInfos {
+	var children InterfaceInfos
+	for _, dev := range s {
+		if dev.ParentInterfaceName == parentName {
+			children = append(children, dev)
+		}
+	}
+	return children
+}
+
+type interfaceInfoByDeviceIndexSlice []InterfaceInfo
+
+func (s interfaceInfoByDeviceIndexSlice) Len() int      { return len(s) }
+func (s interfaceInfoByDeviceIndexSlice) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s interfaceInfoByDeviceIndexSlice) Less(i, j int) bool {
+	dev1 := s[i]
+	dev2 := s[j]
+	return dev1.DeviceIndex < dev2.DeviceIndex
+}
+
+// SortInterfaceInfo sorts a slice of InterfaceInfo on
+// DeviceIndex in ascending order.
+func SortInterfaceInfo(interfaces []InterfaceInfo) {
+	sort.Sort(interfaceInfoByDeviceIndexSlice(interfaces))
```

Comment:
Why not just use [sort.Slice](https://godoc.org/sort#Slice) and pass the predicate as a closure instead of defining a new type?